README.md
# Asterisk REST Interface (ARI) Application Client

forked from https://bitbucket.org/ngvoice/asterisk-ari-client/

A client implementation of the Asterisk REST Interface and simple Stasis
application development library.

The idea is to make ARI calls safe and easy. Therefore, we wanted to get rid of
JSON parsing in our application code. Instead, we aim to make it as easy as possible
for anyone to talk to ARI without worrying about the implementation of a client stub.

![tests](https://github.com/opiy-org/asterisk-ari-php/actions/workflows/main.yml/badge.svg)

![Licence](https://img.shields.io/badge/licence-MIT-blue.svg)

![Diagram of the ARI communication](image/AriClientSketch.png)

## Prerequisites

Download and install composer from the following link:

https://getcomposer.org/download/

## Installing

##### Composer

Please run the following command to add the library to your project:

PHP 7.4
`composer require opiy-org/asterisk-ari-php` 

PHP 8.0|8.1
`composer require opiy-org/asterisk-ari-php:2` 

While installing, you might run into composer errors concerning missing PHP extensions.
There are several ways to install them, depending on your operating system.
In some very unlikely cases, you might need to install `php-dev` first and then install
and enable the extension via PECL. But that is generally not required.

##### Asterisk

You will have to start an Asterisk instance and configure it in order to use ARI.
The official Asterisk documentation shows you how to do so. 

https://wiki.asterisk.org/wiki/display/AST/Asterisk+Configuration+for+ARI

Alternatively, use our Dockerfile to fire up Asterisk ([See Deployment](#deployment)).

## Examples

#### REST Clients

Talk to the Asterisk REST Interface through the given REST clients.
All requests and responses are objects and easy to understand.

The following example originates a call using the Channels resource:

    <?php
    
    declare(strict_types=1);
    
    use OpiyOrg\AriClient\Exception\AsteriskRestInterfaceException;
    use OpiyOrg\AriClient\Client\Rest\Settings as AriRestClientSettings;
    use OpiyOrg\AriClient\Client\Rest\Resource\Channels as AriChannelsRestResourceClient;
    
    require_once __DIR__ . '/vendor/autoload.php';
    
    // Of course inject your own REST client settings here.
    $ariChannelsRestResourceClient = new AriChannelsRestResourceClient(
        new AriRestClientSettings('login', 'password', '127.0.0.1', 8088, 'ari-php')
    );
    
    try {
        // Call the specified number
        $originatedChannel = $ariChannelsRestResourceClient->originate(
            'PJSIP/+4940123456789',
            [
                'app' => 'MyExampleStasisApp'
            ]
        );
    } catch (AsteriskRestInterfaceException $e) {
        printf("Error occurred: '%s'\n", $e->getMessage());
    }
    
    printf("The originated channel has the ID '%s'\n", $originatedChannel->getId());

#### Web socket client

Connects to Asterisk and subscribes to a Stasis application. The following example shows
how to define your Stasis application and how to handle specific incoming events, which
are emitted by a channel that is part of the application context.

In this example, we are handling a `StasisStart` event:
    
    <?php
    
    declare(strict_types=1);
    
    // TODO: Change to your own project namespace.
    namespace My\Own\Project\Namespace;
        
    use OpiyOrg\AriClient\StasisApplicationInterface;
    use OpiyOrg\AriClient\Model\Message\Event\StasisStart;
    
    /**
     * Write your own Stasis application class that must implement the
     * StasisApplicationInterface.
     *
     * This application will register automatically in Asterisk as soon
     * as you start a WebSocketClient (@see example/my_example_stasis_app_worker.php).
     */
    class MyExampleStasisApp implements StasisApplicationInterface
    {
        /**
         * To declare an ARI event handler function, name it after
         * the occurring Asterisk event you want to handle and add
         * the prefix 'onAriEvent'. The only parameter is the object
         * representation of the event, provided by this library.
         * The function MUST also be public and non-static.
         *
         * Of course you can define any other functions within this class
         * that do not handle incoming ARI events (leave out the prefix ;-)).
         * Think of your Stasis application class as container for your
         * application logic.
         *
         * The StasisStart event for example is triggered for
         * channels when they enter your application.
         *
         * @param StasisStart $stasisStart The Asterisk event,
         * telling you that a channel has entered your application.
         */
        public function onAriEventStasisStart(StasisStart $stasisStart): void
        {
            printf(
                'This is the channel's StasisStart event '
                . "handler triggered by channel '%s' :-)\n",
                $stasisStart->getChannel()->getId()
            );
        }
    }

Write a PHP script to start your WebSocketClient worker process.
This is a blocking process! For production, you should use a process manager to run it in
the background. We recommend [supervisor](http://supervisord.org/) for Linux.

    <?php

    declare(strict_types=1);

    use OpiyOrg\AriClient\Client\WebSocket\Factory as AriWebSocketClientFactory;    
    use OpiyOrg\AriClient\Client\WebSocket\Settings as AriWebSocketClientSettings;
    
    require_once __DIR__ . '/vendor/autoload.php';
    require_once __DIR__ . '/MyExampleStasisApp.php';
    
    /*
     * Initialize an ARI web socket client to
     * listen for incoming Asterisk events.
     *
     * Of course inject your own web socket settings here.
     */
    $ariWebSocketClient = AriWebSocketClientFactory::create(
        new AriWebSocketClientSettings('asterisk', 'asterisk'),
        new MyExampleStasisApp()
    );
    
    $ariWebSocketClient->start();

You can find a detailed example with more options in the `example` directory.

## Debug logs

To debug your ARI communication, this client library ships with a simple debug log switch.
Simply en-/disable it as an option in the REST/web socket client settings object. Error
logs will write to STDERR. All other logs will write to STDOUT.

Alternatively, use your own Logger and pass it to the REST client and/or web socket
client constructor.

## Error handler

As described in `example/my_example_stasis_app_worker.php`, you can add a custom error
handler to your application. This is a layer between the logic in your
Stasis application (`e.g. example/MyExampleStasisApp`), and the PHP process error handler.
That means you can decide what to do with a Throwable which is not caught
within your application logic and would normally cause the application (and likely the
whole process) to crash. Why not report it to BugSnag for example?

## Running the tests

### Unit tests

`composer test`

### Coding style tests

For code sniffing

`composer lint `

For static code analysis

`composer sca`

## Deployment

We added a Dockerfile in the `docker/asterisk/` directory, where you can also find some
example configuration files for your own Asterisk instance.

Preferably use the provided Dockerfile in this library to compile your own Asterisk
container.
    
    cd docker/asterisk
    docker build -t --build-arg asterisk_version=WHATEVER_VERSION_YOU_LIKE asterisk:latest .
    docker run -d --name some-asterisk -p 8088:8088 -p 5060:5060 -p 5060:5060/udp asterisk:latest

    !!! PLEASE NOTE !!!
    Compiling Asterisk sometimes is bound to the hardware you are compiling it on.
    Currently, we compile a separate container for every machine we run Asterisk on,
    to make sure it will work.
    Alternatively, you can set generic compiler flags at your own risk.

## Licence

##### MIT

Originally: [ngvoice/asterisk-ari-client](https://bitbucket.org/ngvoice/asterisk-ari-client) (by ng-voice)

## Contributors

Your pull requests are welcome. Please stick to the PSR-12 coding standards if
you want to contribute.

### Maintainer
* [opiy](https://github.com/opiy-org) (opiy@opiy.org)


### Authors
* [Lukas Stermann](https://gitlab.com/Oktavlachs) (lukas@ng-voice.com)
* [Benedikt Vollmerhaus](https://gitlab.com/BVollmerhaus) (Project Review)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.1] - 2023.01.08
add settings property appName

## [2.0.3] - 2022.12.06
add support 8.0, 8.1

## [1.1] - 2021.09.27
fix STDOUT bug

## [1.0.6.1] - 2021.04.22
* Change namespace
* guzzle ^7.0

## [1.0.4] - 2020-01-30

### Fixed
* Default error handler in AbstractWebSocketClient is now non-static

### Removed
* composer.lock from the project, because it must not bind to a specific version

## [1.0.3] - 2020-01-27

### Fixed
* [DialplanCEP](https://wiki.asterisk.org/wiki/display/AST/Asterisk+16+REST+Data+Models#Asterisk16RESTDataModels-DialplanCEP)
properties 'app_data' and 'app_name' are not required although the official
documentation said so
* Unspecific variables names and log information
error handlers must use 'context' as a first parameter in order to
make it reusable throughout the ARI client components (Breaking change!)

## [1.0.2] - 2020-01-24

### Fixed
* Grammar and misleading language in README.md
* Method names and comments in AbstractWebSocketClient and Client\WebSocket\Settings
* Public method names (in classes implementing the StasisApplicationInterface) with
correct ARI event handler prefix but invalid suffix now lead to an
InvalidArgumentException on web socket client connection
* Missing types for method arguments in `Applications::filter()` method
* Missing nullable types for method arguments throughout the project

## [1.0.1] - 2020-01-24

### Fixed
* Grammar and misleading language in README.md

### Added
* Contributors to README

## [1.0.0] - 2020-01-23
**Initial Release**


.phan/config.php
<?php

use Phan\Issue;

/**
 * This configuration file was automatically generated by 'phan --init --init-level=3'
 *
 * TODOs (added by 'phan --init'):
 *
 * - Go through this file and verify that there are no
 * missing/unnecessary files/directories.
 *   (E.g. this only includes direct composer dependencies - You may have to manually
 * add indirect composer dependencies to 'directory_list')
 * - Look at 'plugins' and add or remove plugins if appropriate
 * (see https://github.com/phan/phan/tree/master/.phan/plugins#plugins)
 * - Add global suppressions for pre-existing issues to suppress_issue_types
 * (https://github.com/phan/phan/wiki/Tutorial-for-Analyzing-a-Large-Sloppy-Code-Base)
 *
 * This configuration will be read and overlaid on top of the
 * default configuration. Command line arguments will be applied
 * after this file is read.
 *
 * @see src/Phan/Config.php
 * See Config for all configurable options.
 *
 * A Note About Paths
 * ==================
 *
 * Files referenced from this file should be defined as
 *
 * ```
 *   Config::projectPath('relative_path/to/file')
 * ```
 *
 * where the relative path is relative to the root of the
 * project which is defined as either the working directory
 * of the phan executable or a path passed in via the CLI
 * '-d' flag.
 */
return [

    // Supported values: `'5.6'`, `'7.0'`, `'7.1'`, `'7.2'`, `'7.3'`, `'7.4'`, `null`.
    // If this is set to `null`,
    // then Phan assumes the PHP version which is closest to the minor version
    // of the php executable used to execute Phan.
    //
    // Note that the **only** effect of choosing `'5.6'` is to infer that functions
    // removed in php 7.0 exist.
    // (See `backward_compatibility_checks` for additional options)
    // Automatically inferred from composer.json requirement for "php" of ">=7.3"
    'target_php_version'                                => '7.3',

    // If enabled, missing properties will be created when
    // they are first seen. If false, we'll report an
    // error message if there is an attempt to write
    // to a class property that wasn't explicitly
    // defined.
    'allow_missing_properties'                          => false,

    // If enabled, null can be cast to any type and any
    // type can be cast to null. Setting this to true
    // will cut down on false positives.
    'null_casts_as_any_type'                            => false,

    // If enabled, allow null to be cast as any array-like type.
    //
    // This is an incremental step in migrating away from `null_casts_as_any_type`.
    // If `null_casts_as_any_type` is true, this has no effect.
    'null_casts_as_array'                               => true,

    // If enabled, allow any array-like type to be cast to null.
    // This is an incremental step in migrating away from `null_casts_as_any_type`.
    // If `null_casts_as_any_type` is true, this has no effect.
    'array_casts_as_null'                               => true,

    // If enabled, scalars (int, float, bool, string, null)
    // are treated as if they can cast to each other.
    // This does not affect checks of array keys. See `scalar_array_key_cast`.
    'scalar_implicit_cast'                              => false,

    // If enabled, any scalar array keys (int, string)
    // are treated as if they can cast to each other.
    // E.g. `array<int,stdClass>` can cast to `array<string,stdClass>` and vice versa.
    // Normally, a scalar type such as int could only cast to/from int and mixed.
    'scalar_array_key_cast'                             => true,

    // If this has entries, scalars (int, float, bool, string, null)
    // are allowed to perform the casts listed.
    //
    // E.g. `['int' => ['float', 'string'], 'float' => ['int'],
    // 'string' => ['int'], 'null' => ['string']]`
    // allows casting null to a string, but not vice versa.
    // (subset of `scalar_implicit_cast`)
    'scalar_implicit_partial'                           => [],

    // If enabled, Phan will warn if **any** type in a method invocation's object
    // is definitely not an object,
    // or if **any** type in an invoked expression is not a callable.
    // Setting this to true will introduce numerous false positives
    // (and reveal some bugs).
    'strict_method_checking'                            => false,

    // If enabled, Phan will warn if **any** type in the argument's union type
    // cannot be cast to a type in the parameter's expected union type.
    // Setting this to true will introduce numerous false positives
    // (and reveal some bugs).
    'strict_param_checking'                             => false,

    // If enabled, Phan will warn if **any** type in a returned value's union type
    // cannot be cast to the declared return type.
    // Setting this to true will introduce numerous false positives
    // (and reveal some bugs).
    'strict_return_checking'                            => false,

    // If enabled, Phan will warn if **any** type in a property assignment's union type
    // cannot be cast to a type in the property's declared union type.
    // Setting this to true will introduce numerous false positives
    // (and reveal some bugs).
    'strict_property_checking'                          => false,

    // If true, seemingly undeclared variables in the global
    // scope will be ignored.
    //
    // This is useful for projects with complicated cross-file
    // globals that you have no hope of fixing.
    'ignore_undeclared_variables_in_global_scope'       => true,

    // Set this to false to emit `PhanUndeclaredFunction` issues for internal
    //functions that Phan has signatures for,
    // but aren't available in the codebase, or the internal functions used to run Phan
    // (may lead to false positives if an extension isn't loaded)
    //
    // If this is true(default), then Phan will not warn.
    'ignore_undeclared_functions_with_known_signatures' => true,

    // Backwards Compatibility Checking. This is slow
    // and expensive, but you should consider running
    // it before upgrading your version of PHP to a
    // new version that has backward compatibility
    // breaks.
    //
    // If you are migrating from PHP 5 to PHP 7,
    // you should also look into using
    // [php7cc (no longer maintained)](https://github.com/sstalle/php7cc)
    // and [php7mar](https://github.com/Alexia/php7mar),
    // which have different backwards compatibility checks.
    'backward_compatibility_checks'                     => false,

    // If true, check to make sure the return type declared
    // in the doc-block (if any) matches the return type
    // declared in the method signature.
    'check_docblock_signature_return_type_match'        => false,

    // If true, make narrowed types from phpdoc params override
    // the real types from the signature, when real types exist.
    // (E.g. allows specifying desired lists of subclasses,
    //  or to indicate a preference for non-nullable types over nullable types)
    //
    // Affects analysis of the body of the method
    // and the param types passed in by callers.
    //
    // (*Requires `check_docblock_signature_param_type_match` to be true*)
    'prefer_narrowed_phpdoc_param_type'                 => true,

    // (*Requires `check_docblock_signature_return_type_match` to be true*)
    //
    // If true, make narrowed types from phpdoc returns override
    // the real types from the signature, when real types exist.
    //
    // (E.g. allows specifying desired lists of subclasses,
    // or to indicate a preference for non-nullable types over nullable types)
    //
    // This setting affects the analysis of return statements in
    // the body of the method and the return types passed in by callers.
    'prefer_narrowed_phpdoc_return_type'                => true,

    // If enabled, check all methods that override a
    // parent method to make sure its signature is
    // compatible with the parent's.
    //
    // This check can add quite a bit of time to the analysis.
    //
    // This will also check if final methods are overridden, etc.
    'analyze_signature_compatibility'                   => true,

    // This setting maps case-insensitive strings to union types.
    //
    // This is useful if a project uses phpdoc that differs from the phpdoc2 standard.
    //
    // If the corresponding value is the empty string,
    // then Phan will ignore that union type
    // (E.g. can ignore 'the' in `@return the value`)
    //
    // If the corresponding value is not empty,
    // then Phan will act as though it saw the corresponding UnionTypes(s)
    // when the keys show up in a UnionType of
    // `@param`, `@return`, `@var`, `@property`, etc.
    //
    // This matches the **entire string**, not parts of the string.
    // (E.g. `@return the|null` will still look for a class with the
    // name `the`, but `@return the` will be ignored with the below setting)
    //
    // (These are not aliases, this setting is ignored outside of doc comments).
    // (Phan does not check if classes with these names exist)
    //
    // Example setting: `['unknown' => '', 'number' => 'int|float',
    // 'char' => 'string', 'long' => 'int', 'the' => '']`
    'phpdoc_type_mapping'                               => [],

    // Set to true in order to attempt to detect dead
    // (unreferenced) code. Keep in mind that the
    // results will only be a guess given that classes,
    // properties, constants and methods can be referenced
    // as variables (like `$class->$property` or
    // `$class->$method()`) in ways that we're unable
    // to make sense of.
    'dead_code_detection'                               => false,

    // Set to true in order to attempt to detect unused variables.
    // `dead_code_detection` will also enable unused variable detection.
    //
    // This has a few known false positives, e.g. for loops or branches.
    'unused_variable_detection'                         => false,

    // Set to true in order to attempt to detect redundant and impossible conditions.
    //
    // This has some false positives involving loops,
    // variables set in branches of loops, and global variables.
    'redundant_condition_detection'                     => false,

    // If true, this runs a quick version of checks that takes less
    // time at the cost of not running as thorough
    // of an analysis. You should consider setting this
    // to true only when you wish you had more **undiagnosed** issues
    // to fix in your code base.
    //
    // In quick-mode the scanner doesn't rescan a function
    // or a method's code block every time a call is seen.
    // This means that the problem here won't be detected:
    //
    // ```php
    // <?php
    // function test($arg):int {
    //     return $arg;
    // }
    // test("abc");
    // ```
    //
    // This would normally generate:
    //
    // ```
    // test.php:3 PhanTypeMismatchReturn Returning type string but test()
    // is declared to return int
    // ```
    //
    // The initial scan of the function's code block has no
    // type information for `$arg`. It isn't until we see
    // the call and rescan `test()`'s code block that we can
    // detect that it is actually returning the passed in
    // `string` instead of an `int` as declared.
    'quick_mode'                                        => false,

    // If true, then before analysis, try to simplify AST into a form
    // which improves Phan's type inference in edge cases.
    //
    // This may conflict with `dead_code_detection`.
    // When this is true, this slows down analysis slightly.
    //
    // E.g. rewrites `if ($a = value() && $a > 0) {...}`
    // into `$a = value(); if ($a) { if ($a > 0) {...}}`
    'simplify_ast'                                      => true,

    // Enable or disable support for generic templated
    // class types.
    'generic_types_enabled'                             => true,

    // Override to hardcode existence and types of (non-builtin)
    // globals in the global scope.
    // Class names should be prefixed with `\`.
    //
    // (E.g. `['_FOO' => '\FooClass', 'page' => '\PageClass', 'userId' => 'int']`)
    'globals_type_map'                                  => [],

    // The minimum severity level to report on. This can be
    // set to `Issue::SEVERITY_LOW`, `Issue::SEVERITY_NORMAL` or
    // `Issue::SEVERITY_CRITICAL`. Setting it to only
    // critical issues is a good place to start on a big
    // sloppy mature code base.
    'minimum_severity'                                  => Issue::SEVERITY_LOW,

    // Add any issue types (such as `'PhanUndeclaredMethod'`)
    // to this black-list to inhibit them from being reported.
    'suppress_issue_types'                              => [],

    // A regular expression to match files to be excluded
    // from parsing and analysis and will not be read at all.
    //
    // This is useful for excluding groups of test or example
    // directories/files, unanalyzable files, or files that
    // can't be removed for whatever reason.
    // (e.g. `'@Test\.php$@'`, or `'@vendor/.*/(tests|Tests)/@'`)
    'exclude_file_regex'                                => '@^vendor/.*/(tests?|Tests?)/@',

    // A file list that defines files that will be excluded
    // from parsing and analysis and will not be read at all.
    //
    // This is useful for excluding hopelessly unanalyzable
    // files that can't be removed for whatever reason.
    'exclude_file_list'                                 => [],

    // A directory list that defines files that will be excluded
    // from static analysis, but whose class and method
    // information should be included.
    //
    // Generally, you'll want to include the directories for
    // third-party code (such as "vendor/") in this list.
    //
    // n.b.: If you'd like to parse but not analyze 3rd
    //       party code, directories containing that code
    //       should be added to the `directory_list` as well as
    //       to `exclude_analysis_directory_list`.
    'exclude_analysis_directory_list'                   => [
        'vendor/',
    ],

    // Enable this to enable checks of require/include statements
    // referring to valid paths.
    'enable_include_path_checks'                        => true,

    // The number of processes to fork off during the analysis
    // phase.
    'processes'                                         => 1,

    // List of case-insensitive file extensions supported by Phan.
    // (e.g. `['php', 'html', 'htm']`)
    'analyzed_file_extensions'                          => [
        'php',
    ],

    // You can put paths to stubs of internal extensions in this config option.
    // If the corresponding extension is **not** loaded,
    // then Phan will use the stubs instead.
    // Phan will continue using its detailed type annotations,
    // but load the constants, classes, functions, and classes
    // (and their Reflection types)
    // from these stub files (doubling as valid php files).
    // Use a different extension from php to avoid accidentally loading these.
    // The `tools/make_stubs` script can be used to generate your
    // own stubs (compatible with php 7.0+ right now)
    //
    // (e.g. `['xdebug' => '.phan/internal_stubs/xdebug.phan_php']`)
    'autoload_internal_extension_signatures'            => [],

    // A list of plugin files to execute.
    //
    // Plugins which are bundled with Phan can be added here by providing
    // their name (e.g. `'AlwaysReturnPlugin'`)
    //
    // Documentation about available bundled plugins can be found
    // [here](https://github.com/phan/phan/tree/master/.phan/plugins).
    //
    // Alternately, you can pass in the full path to a PHP file with
    // the plugin's implementation
    // (e.g. `'vendor/phan/phan/.phan/plugins/AlwaysReturnPlugin.php'`)
    'plugins'                                           => [
        'AlwaysReturnPlugin',
        'PregRegexCheckerPlugin',
        'UnreachableCodePlugin',
    ],

    // A list of directories that should be parsed for class and
    // method information. After excluding the directories
    // defined in `exclude_analysis_directory_list`, the remaining
    // files will be statically analyzed for errors.
    //
    // Thus, both first-party and third-party code being used by
    // your application should be included in this list.
    'directory_list'                                    => [
        'src',
        'vendor/dg/bypass-finals/src',
        'vendor/guzzlehttp/guzzle/src',
        'vendor/monolog/monolog/src/Monolog',
        'vendor/nekland/woketo/src',
        'vendor/netresearch/jsonmapper/src',
        'vendor/phan/phan/src/Phan',
        'vendor/phpunit/php-code-coverage/src',
        'vendor/phpunit/phpunit/src',
        'vendor/symfony/yaml',
    ],

    // A list of individual files to include in analysis
    // with a path relative to the root directory of the
    // project.
    'file_list'                                         => [],
];
